Typeracer/src/components/Type.js

```diff
@@ -39,86 +85,282 @@ function Type() {
+                setStatisticsData(prevData => ({
+                    ...prevData,
+                    TypedAmountOfCharacters: prevData.TypedAmountOfCharacters + 1
+                }));
             }
+        } else{
+            event.preventDefault();
         }
-        
-        if (currentIndex + 1 === typingText.length - 1 && inputCharacter === typingText[currentIndex]) { // stops the timer when the text is finished and the last character is typed
+
+        // Using newCurrentIndex for completion check
+        if (!isComplete && newCurrentIndex >= typingText.length) {  // Text completed
             clearInterval(intervalRef.current);
+            const finishTime = Date.now();
+
+            // Collecting updated values
+            const newLocalFinishTime = new Date(finishTime);
+            const newCompletionTime = finishTime - startTime;
+
+            // Preparing the typingData array
+            const typingData = wordsInfoRef.current.map(wordInfo => ({
+                Word: wordInfo.word,
+                BeginningTimestampWord: wordInfo.startTime ? formatDateTime(new Date(wordInfo.startTime)) : null,
+                EndingTimestampWord: wordInfo.endTime ? formatDateTime(new Date(wordInfo.endTime)) : null,
+                AmountOfMistakesInWord: wordInfo.mistakes
+            }));
+
+            // Calculating the total number of wrongful characters
+            const totalMistakes = typingData.reduce((acc, word) => acc + word.AmountOfMistakesInWord, 0);
+
+            // Recalculating TypedAmountOfWords
+            const newTypedAmountOfWords = typingData.filter(td => td.EndingTimestampWord !== null).length;
+
+            // Collecting all updated statistics data
+            const updatedStatisticsData = {
+                ...statisticsData,
+                LocalFinishTime: newLocalFinishTime,
+                CompletionTime: newCompletionTime,
+                TypedAmountOfWords: newTypedAmountOfWords,
+                TypedAmountOfCharacters: currentIndex, // Using currentIndex as the total typed characters
+                NumberOfWrongfulCharacters: totalMistakes,
+                TypingData: typingData
+            };
+
+            // Updating the statisticsData state
+            setStatisticsData(updatedStatisticsData);
+
+            // Assembling the data to send
+            const dataToSend = {
+                CompletionTime: updatedStatisticsData.CompletionTime,
+                LocalStartTime: updatedStatisticsData.LocalStartTime ? formatDateTime(updatedStatisticsData.LocalStartTime) : null,
+                LocalFinishTime: updatedStatisticsData.LocalFinishTime ? formatDateTime(updatedStatisticsData.LocalFinishTime) : null,
+                TotalAmountOfWords: updatedStatisticsData.TotalAmountOfWords,
+                TotalAmountOfCharacters: updatedStatisticsData.TotalAmountOfCharacters,
+                TypedAmountOfWords: updatedStatisticsData.TypedAmountOfWords,
```

Comment:
why is this data being stored twice??

Reply:
Did You mean **TypedAmountOfWords** and **TotalAmountOfWords**? They were in StatisticsInfoModel.cs in the first place

Reply:
omg... sorry i thought it was the same name (total and typed), sleepy head. I forgot the model structure that these are two different stored values. All good. Sorry.